src/vstest.console/InProcessVsTestConsoleWrapper.cs

```diff
@@ -0,0 +1,1101 @@
+
+            testRunEventsHandler.HandleLogMessage(TestMessageLevel.Error, ex.ToString());
+            testRunEventsHandler.HandleTestRunComplete(testRunCompleteArgs, null, null, null);
+        }
+
+        _testPlatformEventSource.TranslationLayerExecutionStop();
+    }
+
+    /// <inheritdoc/>
+    public void RunTestsWithCustomTestHost(
+        IEnumerable<string> sources,
+        string? runSettings,
+        ITestRunEventsHandler testRunEventsHandler,
+        ITestHostLauncher? customTestHostLauncher)
+    {
+        RunTestsWithCustomTestHost(
+            sources,
+            runSettings,
+            options: null,
+            testRunEventsHandler,
+            customTestHostLauncher);
+    }
+
+    /// <inheritdoc/>
+    public void RunTestsWithCustomTestHost(
+        IEnumerable<string> sources,
+        string? runSettings,
+        TestPlatformOptions? options,
+        ITestRunEventsHandler testRunEventsHandler,
+        ITestHostLauncher? customTestHostLauncher)
+    {
+        RunTestsWithCustomTestHost(
+            sources,
+            runSettings,
+            options,
+            testSessionInfo: null,
+            testRunEventsHandler,
+            customTestHostLauncher);
+    }
+
+    /// <inheritdoc/>
+    public void RunTestsWithCustomTestHost(
+        IEnumerable<string> sources,
+        string? runSettings,
+        TestPlatformOptions? options,
+        TestSessionInfo? testSessionInfo,
+        ITestRunEventsHandler testRunEventsHandler,
+        ITestHostLauncher? customTestHostLauncher)
+    {
+        var sourceList = sources.ToList();
+        _testPlatformEventSource.TranslationLayerExecutionStart(
+            1,
+            sourceList.Count,
+            0,
+            runSettings ?? string.Empty);
+
+        try
+        {
+            TestRequestManager?.ResetOptions();
+
+            var shouldLaunchTesthost = true;
+
+            // We must avoid re-launching the test host if the test run payload already
+            // contains test session info. Test session info being present is an indicative
+            // of an already running test host spawned by a start test session call.
+            var customLauncher =
+                shouldLaunchTesthost && testSessionInfo == null
+                    ? customTestHostLauncher
+                    : null;
+
+            var testRunPayload = new TestRunRequestPayload
+            {
+                Sources = sourceList,
+                RunSettings = runSettings,
+                DebuggingEnabled = (customLauncher?.IsDebug == true),
+                TestPlatformOptions = options,
+                TestSessionInfo = testSessionInfo
+            };
+
+            // TODO: this will need to be via an adapter instead. Because we will get a "live"
+            // implementation rather than our own implementation that we normally have in design
+            // mode client. It probably will even throw on this cast.
```

Comment:
what does it mean?

Reply:
I think what this means is that `VsTestConsoleWrapper.cs` gets passed an `ITestHostLauncher` variant that doesn't make its way into `DesignModeClient`, instead we use a proxy of our own design and calls to the proxy are then sent back to the `VsTestConsoleWrapper` via messages. `InProcessVsTestConsoleWrapper` OTOH will use the `ITestHostLauncher` that gets passed in directly and this has some failure potential, but for now it's ok since only VS will adopt this interface in the beginning.

@nohwnd, is this correct ?